### ISSUE TITLE
[ng9] Expression has changed after it was checked. Previous value: 'true'. Current value: 'false'.

### ISSUE THREAD

**Issue by indraraj26** [angular/angular#35817]
# 🐞 bug report

### Affected Package

 The issue is caused by package @angular/....

### Is this a regression?

 Yes, the previous version in which this bug was not present was: ng8

### Description
I am getting below error in console while adding and removing validators.

code: 
```
 handleValidation(addValidationField, removeValidationField) {
    if (addValidationField) {
      let control: FormControl;
      addValidationField.forEach(item => {
        control = this.myForm.get(item) as FormControl;
        control.markAsUntouched();
        control.markAsPristine();
        control.setValidators(Validators.required);
      });
    }
    if (removeValidationField) {
      let control: FormControl;
      removeValidationField.forEach(item => {
        control = this.myForm.get(item) as FormControl;
        control.markAsUntouched();
        control.markAsPristine();
        control.clearValidators();
        control.patchValue("");
      });
    }
    this.myForm.updateValueAndValidity();
  }
```
![image](https://user-images.githubusercontent.com/23274631/75748627-f4d5cb80-5d45-11ea-8840-4d29048d2429.png)

**STEP TO REPRODUCE:**  
Step 1: Select status from dropdown
Step 2: fill anything in below fields
Step 3: change status from dropdown
Step 4: fill anything in below fields

You will get below error in console

## 🔬 Minimal Reproduction

Repo: https://github.com/indraraj26/ng9-barrel/tree/changeRef-issue
Branch: changeRef-issue

## 🔥 Exception or Error
<pre><code>

core.js:5873 ERROR Error: ExpressionChangedAfterItHasBeenCheckedError: Expression has changed after it was checked. Previous value: 'true'. Current value: 'false'.
    at throwErrorIfNoChangesMode (vendor.js:14202)
    at bindingUpdated (vendor.js:24708)
    at interpolation1 (vendor.js:24852)
    at ɵɵtextInterpolate1 (vendor.js:29126)
    at Module.ɵɵtextInterpolate (vendor.js:29097)
    at AppComponent_Template (main.js:225)
    at executeTemplate (vendor.js:17342)
    at refreshView (vendor.js:17189)
    at refreshComponent (vendor.js:18629)
    at refreshChildComponents (vendor.js:16920)
</code></pre>

## 🌍  Your Environment

**Angular Version:**
<pre><code>

ng --version

     _                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/

Angular CLI: 9.0.4
Node: 10.16.0
OS: win32 x64

Angular: 9.0.4
... animations, cli, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router, service-worker
Ivy Workspace: Yes

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.900.4
@angular-devkit/build-angular     0.900.4
@angular-devkit/build-optimizer   0.900.4
@angular-devkit/build-webpack     0.900.4
@angular-devkit/core              9.0.4
@angular-devkit/schematics        9.0.4
@ngtools/webpack                  9.0.4
@schematics/angular               9.0.4
@schematics/update                0.900.4
rxjs                              6.5.4
typescript                        3.7.5
webpack                           4.41.2
</code></pre>

**Anything else relevant?**

**Comment by indraraj26** [comment]
Hi, i got to know this is happening only when i am having this in template.

```
 <pre>{{myForm.valid}}</pre>
    <pre>{{myForm?.value | json}}</pre>
```

Can someone provide reason why is that happening?

**Comment by pkozlowski-opensource** [comment]
@indraraj26 it would help a lot if you could narrow down the scenario to the absolute minimum and reproduce it using https://ng-run.com/ or a similar tool. There is quite a bit going on in your repo and it is not clear what is the root cause...

**Comment by indraraj26** [comment]
Hi @pkozlowski-opensource , 

Here you go :  https://ng-run.com/edit/qDLp6mILN79SXYwsHOUX?open=app%2Fapp.component.html

STEP TO REPRODUCE:
Step 1: Select status from dropdown
Step 2: fill anything in below fields
Step 3: change status from dropdown
Step 4: fill anything in below fields

Thank you.

> i got to know this is happening only when i am having this in template.
> 
> ```
>  <pre>{{myForm.valid}}</pre>
>     <pre>{{myForm?.value | json}}</pre>
> ```
> 
> Can someone provide reason why is that happening?

**Comment by nejat-njonjo** [comment]
I don't get it till now whether to treat it as a warning or an error. Because it is not a bug. 
When the value is checked, Angular keeps them in oldValues during the first check, when second check comes and the value you're assigning is different from the one Angular kept, during first check, that's when this error comes in. Try refactoring your code.

**Comment by nejat-njonjo** [comment]
Also, I think Angular Team should explain a bit more on this problem because it looks like, most of us Angular Devs don't understand this error.

**Comment by indraraj26** [comment]
Hi @pkozlowski-opensource , 

Any update?

**Comment by hemachandran85** [comment]
I am not sure whether Angular team planned for fixes for this issue. For time being , you can suppress the issue by calling the detectChanges() in ChangeDetectorRef class.

Refer : https://angular.io/api/core/ChangeDetectorRef#changedetectorref

**Comment by atscott** [comment]
Hi all, I can understand the frustration with this error, both because it's not obvious what it means and even when you do know, it's not always clear why it's happening and what to do about it. 

TL;DR: This is working as expected and you can resolve the issue by adding 
`control.updateValueAndValidity` to your `forEach` functions and you can remove `this.myForm.updateValueAndValidity();`. 
___
**Additional info**
This is because `updateValueAndValidity` update parents as well, but not children (so `control.updateValueAndValidity` will update the form)
https://github.com/angular/angular/blob/7acd33007d9bd30288a76fa2a3b6fe5948143d9f/packages/forms/src/model.ts#L664

The details of why you encounter this error are a little complicated. First, I'll address what this error means. You can think of Angular's change detection/rendering as a linear series of instructions. When something in a later step updates a value that is rendered in a step that was already processed, we have a problem. It means that because we already processed that node, your newly updated value won't be displayed on that page (until we process the node again in the next change detection). 

The `ExpressionChanged...` error is a check that's run in development mode only. Each change detection cycle actually does two passes. First, the regular run. Next (and only in dev mode), we run a second pass (`checkNoChanges` mode) and compare the values in the application to the values in the DOM. If any of them differ, we know that a change was made to the application's value *after* we had already rendered it. This is where we throw `ExpressionChangedAfterItHasBeenCheckedError`. 

Now it may be worth noting here that specifically calling `changeDetectorRef.detectChanges` triggers an _immediate_ and local change detection/reprocessing of your current component. This means that when the `checkNoChanges` pass happens, your application values will match what's in the DOM. Be careful with `detectChanges` though. Unless you know exactly what you're doing, you'll be triggering additional change detection operations and this could greatly affect performance.

Back to the specific problem you encountered. When you update the validators on the form controls but don't update their validity, the controls are still valid. However, they will become invalid as soon as the template executes the lines to hook up the `formControlName` directive (note that `this.form` below is the `FormControl`, not the `FormGroup`)
https://github.com/angular/angular/blob/7acd33007d9bd30288a76fa2a3b6fe5948143d9f/packages/forms/src/directives/reactive_directives/form_control_directive.ts#L126-L130
You might think "Well, that's fine. Those lines appear before my text interpolation `{{myForm.valid}}` so the validity should get updated before we get there". Unfortunately, it's not that simple. Because those controls are inside an `*ngIf`, they actually execute _after_ your `{{myForm.valid}}` text interpolation. 
If you're curious about the order of execution details, take a look at [refreshView](https://github.com/angular/angular/blob/7acd33007d9bd30288a76fa2a3b6fe5948143d9f/packages/core/src/render3/instructions/shared.ts#L399). `{{myForm.valid}}` is updated as part of the `executeTemplate` line and the `ngIf` contents are updated later, as part of `refreshDynamicEmbeddedViews`.

**Comment by indraraj26** [comment]
Thank you @atscott for awesome explanation. 💥

**Comment by nejat-njonjo** [comment]
Thank you for your response. @tscott

**Comment by nejat-njonjo** [comment]
Thank you for your response @atscott
